### DISCUSSION TITLE
Combined source. Use positional args and dictionary together

### DISCUSSION THREAD

**Discussion by ViacheslavRud** [axuno/SmartFormat#308] [Q&A] [answered]
I need to format the next string:
`Player {PlayerName} gets {0} coins!`
The player name is from Dictionary and {0} is a positional argument.

I've tried to format arguments first, and values from the dictionary next. But it fails trying to resolve {PlayerName} when I provide positional args. 

What is the best way to format such a case?

**Comment by axunonb** [comment, marked as answer]
One option is to put positional arguments to the end:
```CSharp
var dict = new Dictionary<string, string> { { "PlayerName", "Joe" } };
const string format = "Player {PlayerName} gets {1} coins!";
var formatter = Smart.CreateDefaultSmartFormat();
var result = formatter.Format(format, dict, 1234);
// "Player Joe gets 1234 coins!"
```

**Reply by ViacheslavRud** [reply]
Thank you! Could you please clarify where you get the GetSmartFormatter method? Or it's simply a wrapper on Smart.Format()?
I am concerned about this solution's flexibility, if I can place arguments in any order, like mix values from the dictionary and from positional args.

**Reply by axunonb** [reply]
Ha, sorry, should read `Smart.CreateDefaultSmartFormat()`

**Reply by axunonb** [reply]
> solution's flexibility

You can, but you don't have to. A `ValueTuple` is my favorite argument.

**Reply by ViacheslavRud** [reply]
Doesn't work in the next case:
```
        static void Main(string[] args)
        {
            const string format = "Player {PlayerName} gets {1} coins!";
            string result = FormatWrapper(format, "10");
            Console.WriteLine(result);
        }

        private static string FormatWrapper(string txt, params string[] args)
        {
            var dict = new Dictionary<string, string> { { "PlayerName", "Joe" } };
            var formatter = Smart.CreateDefaultSmartFormat();
            var result = formatter.Format(txt, dict, args);
            return result;
        }
```
The smart format will be used deep in the code, I always get args as array or list, it's not "params" from code.

**Comment by axunonb** [comment]
The question contains the answer already: The second parameter is a list. So, to use elements form this parameter it's necessary to reference the index of the array:
```CSharp
[Test]
public void FormatWrapper()
{
    var args = new[] { "1", "2", "3" };
    const string format = "Player {PlayerName} gets {1[0]} or {1[1]} or {1[2]} coins!";
    var dict = new Dictionary<string, string> { { "PlayerName", "Joe" } };
    var formatter = Smart.CreateDefaultSmartFormat();
    var result = formatter.Format(format, dict, args);
    // "Player Joe gets 1 or 2 or 3 coins!"
}
```
(or you could take the `ListFormatter` to evaluate the elements of `args`)

**Reply by ViacheslavRud** [reply]
Ok, I see. Thank you! It's really rather a list than positional args.
